url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+				if (v1.isBool() && v2.isBool()) {
+					v3 = XValueBool.valueOf(v2.getBool() || v1.getBool());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() | v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator2("|", "__lor__", "__lor2__", "__or__", "__or2__", v2, v1);
+			}
+			break;
+		case LPOW:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isNumber() && v2.isNumber()) {
+				if (v1.isFloat() || v2.isFloat()) {
+					v3 = new XValueFloat(Math.pow(v2.getFloat(), v1.getFloat()));
+				} else {
+					v3 = XValueInt.valueOf((long) Math.pow(v2.getInt(), v1.getInt()));
+				}
+				push(v3);
+			} else {
+				callBiOperator2("**", "__lpow__", "__lpow2__", "__pow__", "__pow2__", v2, v1);
+			}
+			break;
+		case LSHL:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isInt() && v2.isInt()) {
+				v3 = XValueInt.valueOf(v2.getInt() << v1.getInt());
+				push(v3);
+			} else {
+				callBiOperator2("<<", "__lshl__", "__lshl2__", "__shl__", "__shl2__", v2, v1);
+			}
+			break;
+		case LSHR:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isInt() && v2.isInt()) {
+				v3 = XValueInt.valueOf(v2.getInt() >> v1.getInt());
+				push(v3);
+			} else {
+				callBiOperator2(">>", "__lshr__", "__lshr2__", "__shr__", "__shr2__", v2, v1);
+			}
+			break;
+		case LSUB:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isNumber() && v2.isNumber()) {
+				if (v1.isFloat() || v2.isFloat()) {
+					v3 = new XValueFloat(v2.getFloat() - v1.getFloat());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() - v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator2("-", "__lsub__", "__lsub2__", "__sub__", "__sub2__", v2, v1);
+			}
+			break;
+		case LXOR:
+			v1 = pop();
+			v2 = pop();
+			if ((v1.isInt() || v1.isBool()) && (v2.isInt() || v2.isBool())) {
+				if (v1.isBool() && v2.isBool()) {
+					v3 = XValueBool.valueOf(v2.getBool() != v1.getBool());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() ^ v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator2("^", "__lxor__", "__lxor2__", "__xor__", "__xor2__", v2, v1);
+			}
+			break;
+		case MAKE_CLASS:
+			s = callFrame.readStringP();
+			v1 = pop();
+			v2 = rt.alloc(rt.getBaseType(XUtils.TYPE), s, v1);
+			push(v2);
+			break;
+		case MAKE_FUNC: {
+			s = callFrame.readStringP();
+			int index = callFrame.readUShort();
+			int size = callFrame.readUByte();
+			String[] params = new String[size];
+			for (int i = 0; i < size; i++) {
+				params[i] = callFrame.readStringP();
+			}
+			int kwParam = callFrame.readByte();
+			int listParam = callFrame.readByte();
+			int defStart = callFrame.readByte();
+			size = callFrame.readUByte();
+			XClosure[] closures = new XClosure[size];
+			for (int i = 0; i < size; i++) {
+				if (callFrame.readByte() == 0) {
+					i1 = callFrame.readUByte();
+					closures[i] = callFrame.getClosure(i1);
+				} else {
+					i1 = callFrame.readUByte();
+					closures[i] = getClosure(i1);
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1287:1295, 1342:1350.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.